### PR TITLE
Migrating marketplace files to a separate section of calypso

### PR DESCRIPTION
**Pull request by ddc22**
#### Changes proposed in this Pull Request

Initially the plan was to bring in the marketplace project under the domain of plugins. So the whole marketplace file structure  lived under the plugins directory. However as the requirement evolved and understandings of nav unification, and how atomic wordpress.com sites operate this seemed to less sense. This was intensified by the fact that routes with "plugins" in them were being forcefully re-directed to the wp-admin plugins page as part of nav unification. 

https://github.com/Automattic/wp-calypso/blob/b2f80f44c7f28503f0d14bcee4f3929466f65eb4/client/my-sites/controller.js#L255-L273

So it made sense to move marketplace to its own section. Moving the rest of the files will be done in a follow up PR 👍 

Note: Code for `/marketplace/thank-you/:site` and `/marketplace/test/:site` will be added at #53107

#### Testing instructions

1. Create a business/ecommerce site
2. Install a plugin
3. Try the following routes
`marketplace/domain/:site`
`/marketplace/product/setup/:site`
`/marketplace/product/details/wordpress-seo/:site`

4. Relevant marketplace routes should be shown and you should not be force-redirected to atomic plugins page.

### CODE REVIEW

File: client/sections.js
Comment (tyxla) [inline client/sections.js]: I've taken a look at the new marketplace controller, and since we are creating those files from scratch, I don't see a good reason not to move this to `calypso/my-sites/marketplace` right here and right now 😉

File: client/my-sites/plugins/marketplace/index.js
Comment (tyxla) [inline client/my-sites/plugins/marketplace/index.js]: Now that we're directing an entire section here, we need to think about all the possible routes that this section is responsible for. Essentially, this means intercepting routes like `/marketplace` and/or `/marketplace/:site`.

File: client/my-sites/plugins/marketplace/index.js
Comment (tyxla) [inline client/my-sites/plugins/marketplace/index.js]: Many site-related sections include the `site` as an optional argument (meaning, changing `:site` to `:site?`) but then include the `siteSelection` middleware there so they'd ensure that the subsequent middleware will have a site available. We've already included site selection, so we only need to make the site argument optional in order to capture any requests that don't include a site in the route.

File: client/sections.js
Comment (tyxla) [inline client/sections.js:113]: I'd like to double-check and verify that this is the URL structure that we want to go with. Once this is introduced and used for something, it could cause some pain to change 😉

File: client/sections.js
Comment (tyxla) [inline client/sections.js:112]: Is there a more specific name we could call this section? Is it a plugin marketplace, or a domain marketplace? I think it would be useful to reflect that in the naming just to avoid potential confusion with any other marketplaces that exist or will exist in the future 😉

File: client/my-sites/plugins/marketplace/controller.js
Comment (tyxla) [inline client/my-sites/plugins/marketplace/controller.js]: I guess let's add some `@TODO` comments here as well. They often further help express intention.

File: client/my-sites/plugins/marketplace/controller.js
Comment (tyxla) [inline client/my-sites/plugins/marketplace/controller.js]: Feel free to get rid of any commented code here, I believe it brings little to no value.

**Review by tyxla — COMMENTED** [review]
Looks good overall! 😉  I'd suggest moving to `client/my-sites/marketplace` or similar, though - it's a good opportunity since we're introducing all those new files in this PR.

I've also left some thoughts and questions, let me know what you think.

File: client/sections.js
Comment (ddc22) [reply client/sections.js]: Yeah but moving this folder would mean breaking a concurrent change I am doing in these files in another branch. This is why I think it would be easier to move the folders after that change has been merged, which has become quite big as it is :  #53107

File: client/my-sites/plugins/marketplace/index.js
Comment (ddc22) [reply client/my-sites/plugins/marketplace/index.js]: Yeah those routes would potentially include a marketplace library of products in the future, for now it will remain empty. Do you mean that we should forward to a known route if someone tries to visit this route? Not sure how to do that and not sure to where to route to though, open to suggestions.

File: client/sections.js
Comment (ddc22) [reply client/sections.js:113]: We currently have no elaborate design, discussion or concensus on the route structure, can you tell me what kind of risk it is and why it would be a potential pain point to change the route structure in the future? 🤔

File: client/sections.js
Comment (tyxla) [reply client/sections.js]: I don't understand that - the `index.js` and `controller.js` are new files, so why not put them in the right place right now? They could reference files from outside for the time being, as long as we fix that in a subsequent PR.

File: client/my-sites/plugins/marketplace/index.js
Comment (tyxla) [reply client/my-sites/plugins/marketplace/index.js]: We could do a redirection, or we could display an error screen, we're doing both in Calypso. Whatever works best for the user flows you're building.

File: client/sections.js
Comment (tyxla) [reply client/sections.js:113]: The idea is that if we introduce this to the public, and eventually we decide to rename it, we'll have to create redirects, potentially update any links and documentation that references the old links as well. I don't have reservations on the naming specifically, just wanted to get your confirmation that it's been agreed both from a product and marketing perspective that `http://wordpress.com/marketplace/*` will be the final destination of this new section.

File: client/sections.js
Comment (ddc22) [reply client/sections.js:112]: Having the name plugin-marketplace can be confusing since we already have a "plugins" place to add plugins. Also we bundle upsell a domain here. So the generic marketplace is the right term for it, at least right now.

Aspirations for an elaborate marketplace are far into the future and we agreed to not let any nuances of the "marketplace" effect our technical decisions. Basically the instructions were to go with a separate "Yoast Sale" page if required. So we can live with this name for now and allowances and analysis will be made when we truly want this to be a generic marketplace or specialised one. I went with this names so that we are making at least some sort of allowances in our code for it and to make sure the user would not be confused by a route like wordpress.com/yoast.

File: client/my-sites/plugins/marketplace/controller.js
Comment (ddc22) [reply client/my-sites/plugins/marketplace/controller.js]: These will be addressed by my immediate next commit that is why I did not put anything extra :)

File: client/my-sites/plugins/marketplace/controller.js
Comment (ddc22) [reply client/my-sites/plugins/marketplace/controller.js]: This change was done inside an already large PR #53107 as I mentioned above. I cherry picked this change because introducing a section is kind of a big deal.So this code is to be rebased with that PR and the code commented out is required.

File: client/sections.js
Comment (tyxla) [reply client/sections.js:112]: Sounds good to me, and this is a product decision so I'm totally fine with whatever you'd like to go with. Just wanted to get your thoughts on it 👍

File: client/my-sites/plugins/marketplace/controller.js
Comment (tyxla) [reply client/my-sites/plugins/marketplace/controller.js]: Sounds good. Feel free to split to multiple PRs if that helps 😉

**Comment by ddc22** [comment]
@tyxla as I explained this change is part of the change I am working at #53107 so there are changes already in a branch which would result in merge hell if I move this now. After that change is done I will definitely move this folder to the desired location.

File: client/my-sites/plugins/marketplace/controller.js
Comment (tyxla) [reply client/my-sites/plugins/marketplace/controller.js]: Sounds good as long as it's not still commented in 2022 🤣

**Comment by tyxla** [comment]
> @tyxla as I explained this change is part of the change I am working at #53107 so there are changes already in a branch which would result in merge hell if I move this now. After that change is done I will definitely move this folder to the desired location.

Sounds good @jdc91  👍  Let me know if you need a hand with any of that.

File: client/my-sites/plugins/marketplace/index.js
Comment (tyxla) [inline client/my-sites/plugins/marketplace/index.js]: It's worth thinking about what happens in production if someone wants to access `/marketplace` and this is disabled. Perhaps we should just redirect to `wordpress.com`?

File: client/sections.js
Comment (ddc22) [reply client/sections.js]: Oh ok you meant to move those two files only 😄 
In my mind I thought you meant have everything under the correct folder.
Will do that change 👍

File: client/sections.js
Comment (ddc22) [reply client/sections.js:113]: @paulbonahora @JanaMW27 just a small heads up here. The route structure for our marketplace stuff would be 
`http://wordpress.com/marketplace/*`. I know we haven't discussed this but it might be bit of a brittle change so looping you in for thoughts.

File: client/my-sites/plugins/marketplace/controller.js
Comment (ddc22) [reply client/my-sites/plugins/marketplace/controller.js]: 🤣 Hehe hopefully 🤣 
Actually you have my word 💯

File: client/sections.js
Comment (tyxla) [reply client/sections.js]: Yes. 😉 

FWIW I'm totally fine with following up in another PR to move the rest.

File: client/my-sites/plugins/marketplace/index.js
Comment (ddc22) [reply client/my-sites/plugins/marketplace/index.js]: Hmmm I guess so but we can pick this up later 👍 I will tag it as an issue c/Uj01USuX-tr

File: client/my-sites/plugins/marketplace/index.js
Comment (ddc22) [reply client/my-sites/plugins/marketplace/index.js]: Will pick this up later seperately c/Uj01USuX-tr

File: client/my-sites/plugins/marketplace/index.js
Comment (jaswrks) [reply client/my-sites/plugins/marketplace/index.js]: ➕1  on adding a default case handler with a redirection here to avoid broken paths later on. I recommend adding it here and now to avoid the need to track yet another TODO. I'd rather see us implement a short-term solution in code and then iterate on that simple redirect in the future.

Comment (tyxla) [reply client/my-sites/plugins/marketplace/index.js]: Agreed it's a rather minor change. You can use an asterisk as a catch-all when declaring routes, as it's done here: https://github.com/Automattic/wp-calypso/blob/dac30586cd5253c2bf463635e869a364e24bcdcc/client/my-sites/comments/index.js#L47

Comment (ddc22) [reply client/my-sites/plugins/marketplace/index.js]: Hi @tyxla @jaswrks thanks for that nudge, didn't realise it would be this easy :D
Fixed by d03d31c486c6d083d184793edcc2f744ae2603e0

File: client/my-sites/marketplace/controller.js
Comment (southp) [inline client/my-sites/marketplace/controller.js]: Let's keep the capitalization style consistent. By spot-checking a few files like https://github.com/Automattic/wp-calypso/blob/trunk/client/state/index.js#L2, I think this should be `Internal dependencies`

**Review by southp — APPROVED** [review]
Great work @jdc91, and thanks for your invaluable assistance on this, @tyxla . Although we are moving routes to mitigate conflicting to the AT overrides, this reads a lot cleaner to me. 🚢 

@tyxla I'm approving it here since I don't see obvious issues to unblock @jdc91 . If there are further feedbacks for us to address, please still feel free to comment and we will be happy to work on them 🙇🏼

File: client/my-sites/marketplace/index.js
Comment (tyxla) [inline client/my-sites/marketplace/index.js:51]: Feels odd to have to pick a site just to know where to redirect to. the `/home` route already has site selection middleware, so maybe we should let it handle site selection and just stick with redirection here?

**Review by tyxla — COMMENTED** [review]
> @tyxla I'm approving it here since I don't see obvious issues to unblock @jdc91 . If there are further feedbacks for us to address, please still feel free to comment and we will be happy to work on them 🙇🏼 

Sounds good, I've taken another look and everything looks sensible FWIW 👍  I've just left a minor non-critical suggestion.

**Pull request merged**
